src/components/structures/MatrixChat.js

```diff
@@ -592,23 +575,40 @@ module.exports = React.createClass({
         this.scrollStateMap[roomId] = state;
     },
 
-    onLoggedIn: function(credentials) {
+    _doLogin(creds) {
+        Lifecycle.login(creds, {
+            syncTimelineLimit: this.props.config.sync_timeline_limit,
+        });
+    },
+
+    _onHaveCredentials: function(credentials) {
         credentials.guest = Boolean(credentials.guest);
-        console.log("onLoggedIn => %s (guest=%s)", credentials.userId, credentials.guest);
-        MatrixClientPeg.replaceUsingAccessToken(
-            credentials.homeserverUrl, credentials.identityServerUrl,
-            credentials.userId, credentials.accessToken, credentials.guest
-        );
+        Lifecycle.login(credentials);
+    },
+
+    _onLoggedIn: function(credentials) {
+        this.guestCreds = null;
         this.setState({
             screen: undefined,
-            logged_in: true
+            logged_in: true,
         });
-        this.startMatrixClient();
-        this.notifyNewScreen('');
     },
 
-    startMatrixClient: function() {
+    _onLoggedOut: function() {
+        this.notifyNewScreen('login');
+        this.replaceState({
+            logged_in: false,
+            ready: false,
+        });
+    },
+
+    _onWillStartClient() {
         var cli = MatrixClientPeg.get();
+
+        if (this.props.config.sync_timeline_limit) {
+            MatrixClientPeg.opts.initialSyncLimit = this.props.config.sync_timeline_limit;
```

Comment:
Yeah, good point - in the case of options there's no reason it needs to be exactly at this point.